Use PatchStatus to update node status in kubelet.

Random-Liu opened the pull request:
Fixes https://github.com/kubernetes/kubernetes/issues/37771.

This PR changes kubelet to update node status with `PatchStatus`.

@caesarxuchao @ymqytw told me that there is a limitation of current `CreateTwoWayMergePatch`, it doesn't support primitive type slice which uses strategic merge.
* I checked the node status, the only primitive type slices in NodeStatus are as follows, they are not using strategic merge:
  * [`ContainerImage.Names`](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L2963)
  * [`VolumesInUse`](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L2909)
* Volume package is already [using `CreateStrategicMergePath` to generate node status update patch](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go#L111), and till now everything is fine. 

@yujuhong @dchen1107 
/cc @kubernetes/sig-node

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37871)
<!-- Reviewable:end -->

Random-Liu committed:
Use PatchStatus to update node status in kubelet.

k8s-oncall commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37871)
<!-- Reviewable:end -->

ymqytw commented:
@Random-Liu PATCH doesn't work for MERGING list of primitives, such as [finalizers](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L205). Merging list should specify `patchStrategy:"merge"`. Other list of primitives use replace patchStrategy for default.

>ContainerImage.Names
VolumesInUse

The 2 you mentioned above are using replace patchStrategy. I believe PATCH should work for them by replacing the whole list.

Random-Liu commented:
@ymqytw Thanks for confirmation! :)

Random-Liu commented:
Will update the PR to fix the unit test.

Random-Liu commented:
@yujuhong Fixed the unit test. Please take a look. :)

rmmh commented:
@k8s-bot gci gce e2e test this (job was accidentally deleted)

yujuhong commented:
@caesarxuchao, could you review the status update operation? Thanks!

caesarxuchao commented on pkg/util/node/node.go:
```diff
@@ -153,3 +154,27 @@ func SetNodeCondition(c clientset.Interface, node types.NodeName, condition v1.N
+		return nil, fmt.Errorf("failed to create patch for node %q: %v", nodeName, err)
+	}
+
+	updatedNode, err := c.Core().Nodes().PatchStatus(string(nodeName), patchBytes)
```
Could you switch to use the regular Patch() method? It supports specifying subresources now. We should deprecate the special PatchStatus() method.

Random-Liu replied:
So now `Patch` can update status, too?

caesarxuchao replied:
Yeah, you need to set the `subreources`.
```
Patch(name string, pt api.PatchType, data []byte, subresources ...string) (result *v1.Node, err error)
```

Random-Liu replied:
Done

Random-Liu commented on pkg/util/node/node.go:
```diff
@@ -153,3 +154,27 @@ func SetNodeCondition(c clientset.Interface, node types.NodeName, condition v1.N
 }
+
+// UpdateNodeStatus updates node status with PatchStatus.
+func UpdateNodeStatus(c clientset.Interface, nodeName types.NodeName, oldNode *v1.Node, newNode *v1.Node) (*v1.Node, error) {
```
Ha, I see. Just set the old one to new one. I see.

Random-Liu replied:
Done

Random-Liu commented:
The unit test `TestTryRegisterWithApiServer` failed because we are actually using `UpdateStatus` to update `ObjectMeta`(`Annotation`). (See https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L143-L171)

We could:
1) Explicitly change the function from `UpdateNodeStatus` to `UpdateNode`.
2) Still name the function as `UpdateNodeStatus` and do not reset `ObjectMeta` inside the container.

@caesarxuchao Which do you prefer?

Random-Liu commented:
Offline discussed with @caesarxuchao, only reset `Node.Spec` in `UpdateNodeStatus`, because:
1. Apiserver allows updating objectmeta when updating node status.
2. Some component does use this to update node annotations.

Random-Liu commented:
@caesarxuchao @yujuhong Addressed comments. Let's merge this PR after https://github.com/kubernetes/kubernetes/pull/38375 to avoid conflict during cherry-pick.

Random-Liu commented:
@k8s-bot gce etcd3 e2e test this

Random-Liu commented:
@caesarxuchao PR ready to go.

caesarxuchao commented:
/lgtm
